docs/rfc/0696-p2p-node.adoc

```diff
@@ -0,0 +1,318 @@
+with the `track` operation for remotes in the context of a `Urn`. To enable the
+same approach as for announcements where offline changes to the storage can be
+picked up at any point the node is running and online, this subroutine will
+maintain a persisted view and periodically (potentially provoked by a request
+over the <<API>>) build the delta for new tracked `Urn`s and remotes in their
+context.
+
+Note: A blocking issue for that is
+https://github.com/radicle-dev/radicle-link/issues/141[#141]
+
+==== Tracking
+
+Given the ability to configure a set of urn, peers, both or anything observed
+from connected peers, the node will automatically track and replicate. This
+strategy is currently in use in the seed node and SHALL be preserved.
+
+== API
+
+=== IPC
+
+All communication with daemon processes SHALL occur over UNIX domain sockets in
+`SOCK_STREAM` mode. The socket files MUST be stored in a directory only
+accessible by the logged-in user (typically `$XDG_RUNTIME_DIR`), and have
+permissions `0700` by default. Per-service socket paths are considered "well
+known".
+
+RPC calls over those sockets use <<cbor, CBOR>> for their payload encoding. As
+incremental decoders are not available on all platforms, CBOR-encoded messages
+shall be prepended by their length in bytes, encoded as a 32-bit unsigned
+integer in network byte order.
+
+RPC messages are wrapped in either a `request` or `response` envelope structure
+as defined below:
+
+[source,cddl]
+----
+request = [
+    request-headers,
+    ? payload: bstr,
+]
+response = ok / error
+ok = [
+    response-headers,
+    ? payload: bstr,
+]
+error = [
+    response-headers,
+    code: uint,
+    ? message: tstr,
+]
+request-headers = {
+    ua: client-id,
+    rq: request-id,
+    ? token: token,
+}
+response-headers = {
+    rq: request-id,
+}
+; Unambiguous, human-readable string identifying the client application. Mainly
+; for diagnostic purposes. Example: "radicle-link-cli/v1.2+deaf"
+client-id: tstr .size (4..16)
+; Request identifier, choosen by the client. The responder includes the
+; client-supplied value in the response, enabling request pipelining.
+;
+; Note that streaming / multi-valued responses may include the same id in
+; several response messages.
+request-id: bstr .size (4..16)
+
+; Placeholder for future one-time-token support.
+token: bstr
+----
+
+
+=== Request/Response payloads
+
+There is a supported set of requests a daemon will answer to which are as follows:
+
+[source,cddl]
+----
+request = discover / memebership / stats
+request-discover = {
+  type: "discover",
+  info: connect-info,
+}
+request-memebership = {
+  type: "membership",
+}
+request-stats = {
+  type: "stats",
+}
+
+response = membership / stats
+response-memebership = {
+  active: [peer-id],
+  passive: [peer-id],
+}
+response-stats = {
+  connections-total: uint,
+  connected-peers: [connect-info],
+  memebership-active: uint,
+  memebership-passive: uint,
+}
```

Comment:
```suggestion
response = membership-info / stats
membership-info = {
  active: [* peer-id],
  passive: [* peer-id],
}
stats = {
  "connections-total" ^ => uint,
  "connected-peers" ^ => [* connect-info],
  "membership-active" ^ => uint,
  "membership-passive" ^ => uint,
  * tstr => any
}
```

I'm specifically suggesting to use string keys for the `stats` response, as it's not too important to make this particularly compact, but important to be able to change things later by adding or removing fields. The "cut" (`^ =>`) in combination with `* tstr => any` says that we're free to add more fields to the map (of any type), but the specified keys must be of the given type. Eg. if there was a value '"connections-total": "NaN"', `* tstr => any` is not allowed to match.

Reply:
On second pass, I'm not too enthused about handing out IP addresses in the stats. This should be removed. Stats is about numbers.

Reply:
As a counterpoint, I have found this information extremely useful for debugging purposes. It's useful to be able to understand problems with connectivity when you can see what the actual network addresses you are connecting (or not connecting) to are. I think this is important for the peer to peer node because we're no longer in a situation where it's easy to ask users to just invoke Upstream (for example) with some additional logging. They would need to change the logging for the peer to peer node (which they may not even be aware of) and then restart the node. This API means the application can directly provide useful diagnostics to the user.

Reply:
Then go design a debugging facility for authorized personell. We do not conciously expose IP addresses as part of a public interface. There is nothing to discuss about this, unless you are okay with transferring 4% of your yearly revenue to the project legal treasury under GDPR.

Reply:
My angle here was to expose the public introspection methods of `net::peer::Peer` for stats that is `librad::net::protocol::event::downstream::Stats`. This struct has the `conencted_peers` field which is a map of peer_id to `Vec<SocketAddr>`. Given that, would it therefore make sense to remove that information from that struct in the first place. So that the invariant that IP information are not exposed publicly will hold in the future and the generations to follow won't get tempted by the power at their fingertips?

Reply:
> This struct has the conencted_peers field

Wait what? That's a bug then.


I know this is hard to grasp for the code-is-law crowd, but the truth is that law is a Wittgensteinian business. It is actually very simple: collecting IP addresses in secondary storage is not necessary for providing the service of running a radicle link node. It _is_ necessary, however, to ensure the quality of the service under below-optimum network conditions. We cannot expect the average operator of a radicle link node (be it a publicly advertised node or a "private" laptop node) to be aware of any data protection obligations they acquire implicitly by doing so. 

Therefore, the **default** must be data frugality. Only _when you know what you're doing_, and you're doing it _conciously_, you should be able to access data where explicit consent is somewhat difficult to acquire.

**tl;dr** Stats are a thing we'd provide for people to put on a dashboard. It should consist of numbers. Administrative introspection can be a separate channel, which should preferably be disabled by default, and can be enabled when needed.

Reply:
> Wait what? That's a bug then.

https://github.com/radicle-dev/radicle-link/blob/138813dba6d2cf377865ba163049982cb4c03c1c/librad/src/net/protocol/event.rs#L56-L63

Reply:
This is something I built when I first started to support the network diagnostics panel, I wasn't aware of the stance on exposing public IP addresses.

Reply:
I found your name in `git blame` already! **/shakesfist** Seems like I missed that change in code review.

Reply:
:flushed:

Reply:
Reworked the requests, to match the `net::Peer` API where there is a call to get connected peers which returns just the list of peer ids in 57e6c29 - also added an item in #722 to remove the offensive field from `Stats` entirely. Speak up if anyone of you see further need to capture the outcome of this discussion.

@alexjg One thing that is apparent from the exchange in this thread is that we should early on think about introspection for _authorized personnel_ and I was wondering if you have - from the experience building that for the current daemon - a list of requirements, constraints, prior art that could inform that.

Reply:
Yeah there's a number of thoughts I have on this. My plan was to put it into an RFC once the CLI and daemon RFCs are merged (so that we're working from a somewhat stable base). My instinct is that we should not have too many in progress RFCs which relate to each other. Sound sensible or should we just write the RFC?

Reply:
> Sound sensible or should we just write the RFC?

Sounds sensible to me, at the same time I'd read a draft :) But I'm sure you'd also like to concentrate on the COBs too

Reply:
Closing this thread and follow-up work is tracked in #722